Delay AsyncInsertQueue push to avoid some spurious failures and improve code documentation

aiven-sal opened the pull request:
If an exception is thrown after the query is pushed into the queue,
the user may receive and error even if the insert succeeded or if it is
still in the queue.
This would be even more problematic when the fix for #47593 will be
implemented, because, once you give up control of the socket to the bg
thread, the original thread can't send errors to the user anymore.
This commit doesn't completely solve the issue, the callers of executeQuery
should be modified too to avoid throwing issues after the push.
It will be improved more with the fix for #47593 and other future PRs.
But it should be easier to review it, if I submit it separately.

Also improve the description of one of the overloads of executeQuery, which
didn't reflect perfectly its behaviour when dealing with async inserts. 


### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

robot-ch-test-poll commented:
<!-- automatic status comment for PR #50436 from aiven-sal/ClickHouse:aiven-sal/async_insert_eh_pr -->
This is an automated comment for commit 6a5d70c7d22194ae523e0ad1a5da633bc66cc5de with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/50436/6a5d70c7d22194ae523e0ad1a5da633bc66cc5de/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

alexey-milovidov commented on src/Interpreters/AsynchronousInsertQueue.h:
```diff
@@ -30,15 +30,14 @@ class AsynchronousInsertQueue : public WithContext
     };
 
-    PushResult push(ASTPtr query, ContextPtr query_context);
+    PushResult pushCheckOnly(ASTPtr query, ContextPtr query_context, String & bytes);
```
Comments?

alexey-milovidov commented on src/Interpreters/executeQuery.cpp:
```diff
@@ -307,13 +307,14 @@ static void setQuerySpecificSettings(ASTPtr & ast, ContextMutablePtr context)
 }
 
-static std::tuple<ASTPtr, BlockIO> executeQueryImpl(
+static std::tuple<ASTPtr, BlockIO, bool, String> executeQueryImpl(
```
Comments?

alexey-milovidov commented on src/Interpreters/executeQuery.cpp:
```diff
@@ -307,13 +307,14 @@ static void setQuerySpecificSettings(ASTPtr & ast, ContextMutablePtr context)
     QueryProcessingStage::Enum stage,
-    ReadBuffer * istr)
+    ReadBuffer * istr,
+    std::optional<std::future<void>> future = std::nullopt)
```
A comment?

alexey-milovidov commented on src/Interpreters/executeQuery.cpp:
```diff
@@ -612,14 +615,17 @@ static std::tuple<ASTPtr, BlockIO> executeQueryImpl(
                 if (settings.wait_for_async_insert)
                 {
+                    if (!future)
+                        throw Exception(ErrorCodes::LOGICAL_ERROR, "wait_for_async_insert = 1 but not future to wait for");
```
You meant "no future" instead of "not future".

alexey-milovidov commented on src/Interpreters/executeQuery.cpp:
```diff
@@ -1190,6 +1201,21 @@ BlockIO executeQuery(
     }
+    if (async_insert)
+    {
+        auto * queue = context->getAsynchronousInsertQueue();
```
This code looks strange.

alexey-milovidov replied:
Missing comments in `Context.h`, before this line:
```
AsynchronousInsertQueue * getAsynchronousInsertQueue() const;
```

The previous author was wrong, but now we have to fix their faults.

alexey-milovidov commented on src/Interpreters/executeQuery.h:
```diff
@@ -36,6 +36,8 @@ void executeQuery(
 /// Returns a pair of block streams which, when used, will result in query execution.
 /// This means that the caller can to the extent control the query execution pipeline.
+/// Be aware that asynchronous insert queries will be queued for execution by this
+/// function and will be executed regardless of what you do after this function run.
```
This sounds unfriendly.

alexey-milovidov commented:
I've asked the developer of this code, and he answered:

> I’m confused with this PR. I don’t understand what the author wants, but changes look strange to me.

aiven-sal committed:
Delay AsyncInsertQueue push to avoid spurious failures

AsyncInsert are implemented by completely decouple the actual
execution of the query (which happens in a background thread)
from the execution of the code that the rest of CH considers
the query's execution.
This can lead to some idiosyncrasies:

* A user can receive a error for an insert that actually succeeded.
* A user can receive a error for an insert that didn't start yet
  and may or may not succeed.
* A request to kill an insert can complete successfully and the
  user that run the query will receive an error, but query won't
  actually be killed.
* A query prepared using the lower level executeQuery overload, will
  be executed when it is not supposed to.
* Maybe more...

This commit delays the push of the query to the execution query
in an attempt to limit all of the above issues.

alexey-milovidov commented:
@aiven-sal Let's resolve conflicts and fix the style.
